# Conhendo os Dados

In [1]:
import numpy as np 
import pandas as pd

Os dados são obtidos do site [kaggle](https://www.kaggle.com/datasets/surajjha101/cuisine-rating?resource=download), onde se pode obter mais informações sobre a estrutura dos dados.

In [2]:
cuisine_rating = pd.read_csv('cuisine_rating.csv')

In [3]:
cuisine_rating.head()

User ID  Area code            Location  Gender   YOB Marital Status  \
0        1        153  Upper East Side,NY  Female  2006         Single   
1        2        123       St. George,NY  Female  1991        Married   
2        3        122  Upper West Side,NY    Male  1977         Single   
3        4        153  Upper East Side,NY  Female  1956        Married   
4        5        129     Central Park,NY    Male  1997         Single   

       Activity  Budget  Cuisines  Alcohol     Smoker  Food Rating  \
0  Professional       3  Japanese     Never     Never            5   
1       Student       3    Indian     Never  Socially            1   
2       Student       5   Seafood     Often     Often            5   
3  Professional       5  Japanese     Never  Socially            3   
4       Student       4  Filipino  Socially     Never            2   

   Service Rating  Overall Rating Often A S  
0               4             4.5        No  
1               1             1.0        No  
2               5             5.0       Yes  
3               1             2.0        No  
4               4             3.0        No

In [4]:
cuisine_rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   User ID         200 non-null    int64  
 1   Area code       200 non-null    int64  
 2   Location        200 non-null    object 
 3   Gender          200 non-null    object 
 4   YOB             200 non-null    int64  
 5   Marital Status  200 non-null    object 
 6   Activity        200 non-null    object 
 7   Budget          200 non-null    int64  
 8   Cuisines        200 non-null    object 
 9   Alcohol         200 non-null    object 
 10  Smoker          200 non-null    object 
 11  Food Rating     200 non-null    int64  
 12  Service Rating  200 non-null    int64  
 13  Overall Rating  200 non-null    float64
 14  Often A S       200 non-null    object 
dtypes: float64(1), int64(6), object(8)
memory usage: 23.6+ KB


Qual tipo de culinária é mais popular?

In [5]:
cuisine_rating['Cuisines'].value_counts()

Cuisines
Japanese    36
French      34
Filipino    34
Indian      32
Chinese     24
Seafood     22
Italian     18
Name: count, dtype: int64

In [6]:
cuisine_rating[cuisine_rating['Cuisines'] == 'Japanese']['Overall Rating'].mean() # Temos a média das avaliações da Cuzinha Japonesa

np.float64(3.5972222222222223)

# Fazendo Agraupamentos

Alguns momentos temos interesse de sumarizar algumas informações.

O problema caso é identificar qual melhor culinária no Central Park, o que vai indicar o melhor cenário para abrir um novo negocio.

In [7]:
cuisine_rating.describe() # Realiza uma descrição simplificada dos dados

User ID   Area code          YOB      Budget  Food Rating  \
count  200.000000  200.000000   200.000000  200.000000   200.000000   
mean   100.500000  141.060000  1984.830000    3.815000     3.220000   
std     57.879185   26.130257    16.809339    1.056578     1.411226   
min      1.000000  101.000000  1955.000000    1.000000     1.000000   
25%     50.750000  123.000000  1971.000000    3.000000     2.000000   
50%    100.500000  135.000000  1987.000000    4.000000     3.000000   
75%    150.250000  158.000000  2000.000000    5.000000     5.000000   
max    200.000000  199.000000  2009.000000    5.000000     5.000000   

       Service Rating  Overall Rating  
count      200.000000      200.000000  
mean         3.230000        3.225000  
std          1.526022        1.079445  
min          1.000000        1.000000  
25%          2.000000        2.500000  
50%          3.000000        3.000000  
75%          5.000000        4.000000  
max          5.000000        5.000000

In [8]:
cuisine_rating['Location'].value_counts()

Location
St. George,NY         46
Upper East Side,NY    30
Riverdale,NY          28
Central Park,NY       24
China Town, NY        22
Market City, NY       20
Upper West Side,NY    18
Central Park,ny        8
Market City, MY        2
Cedar Hill, NY         2
Name: count, dtype: int64

In [9]:
numeric_cols = cuisine_rating.select_dtypes(include=np.number).columns.to_list()
numeric_cols

['User ID',
 'Area code',
 'YOB',
 'Budget',
 'Food Rating',
 'Service Rating',
 'Overall Rating']

In [10]:
cuisine_rating.groupby(['Location', 'Cuisines'])[numeric_cols].mean()

User ID   Area code          YOB    Budget  \
Location           Cuisines                                                  
Cedar Hill, NY     Indian    141.000000  169.500000  1960.000000  4.000000   
Central Park,NY    Chinese    60.000000  129.000000  1995.000000  4.000000   
                   Filipino   83.666667  144.166667  1990.666667  3.333333   
                   French    116.000000  144.750000  2003.500000  4.000000   
                   Indian     77.000000  131.500000  2001.000000  3.000000   
                   Italian   115.000000  152.500000  1996.000000  4.000000   
                   Japanese   98.000000  138.500000  2002.000000  2.500000   
                   Seafood    80.500000  149.500000  1980.000000  3.000000   
Central Park,ny    Filipino  108.000000  162.500000  1962.000000  3.000000   
                   French     97.000000  117.500000  1976.000000  3.000000   
                   Indian     79.500000  137.000000  1986.000000  4.000000   
China Town, NY     Chinese    96.000000  106.500000  1976.000000  3.000000   
                   Filipino  130.000000  125.000000  1998.000000  1.000000   
                   French    130.500000  149.000000  1965.000000  3.500000   
                   Indian     93.333333  129.333333  1983.666667  4.333333   
                   Seafood   100.250000  135.000000  1991.750000  3.500000   
Market City, MY    Indian     82.000000  156.500000  1971.000000  1.000000   
Market City, NY    Filipino   92.000000  158.000000  1967.000000  5.000000   
                   French    105.000000  156.250000  1975.500000  4.000000   
                   Indian    122.000000  169.000000  1977.500000  3.000000   
                   Japanese  114.000000  152.333333  1981.333333  4.666667   
                   Seafood    77.000000  144.750000  1981.000000  3.000000   
Riverdale,NY       Chinese    77.000000  121.000000  1985.000000  4.500000   
                   Filipino  104.666667  116.666667  1966.333333  4.333333   
                   French    116.000000  150.500000  1975.000000  4.000000   
                   Italian    81.000000  124.333333  1981.000000  3.666667   
                   Japanese  105.750000  135.250000  1994.500000  4.500000   
                   Seafood    67.000000  133.000000  2007.000000  5.000000   
St. George,NY      Chinese    62.500000  124.000000  1984.500000  4.500000   
                   Filipino  121.166667  136.083333  1991.000000  4.000000   
                   French    109.500000  127.875000  1986.750000  4.500000   
                   Indian     96.500000  145.625000  1987.500000  3.375000   
                   Italian   101.000000  155.250000  1975.000000  2.500000   
                   Japanese  107.800000  140.800000  1985.200000  4.400000   
Upper East Side,NY Chinese   120.000000  159.400000  1980.000000  4.000000   
                   Filipino   58.000000  159.500000  2003.000000  3.000000   
                   French    101.666667  145.833333  1982.666667  3.666667   
                   Indian    124.000000  172.500000  1994.000000  3.000000   
                   Italian    93.000000  157.750000  1986.000000  5.000000   
                   Japanese   57.666667  148.666667  1988.666667  3.666667   
Upper West Side,NY Chinese   147.000000  146.000000  2000.000000  4.000000   
                   Filipino  106.000000  145.500000  1991.000000  4.000000   
                   French    136.000000  140.000000  1999.500000  4.000000   
                   Indian     83.000000  141.500000  1995.000000  3.000000   
                   Italian   110.000000  128.500000  1959.000000  5.000000   
                   Japanese  112.000000  123.500000  1955.000000  4.000000   
                   Seafood    75.500000  123.500000  1982.000000  4.000000   

                             Food Rating  Service Rating  Overall Rating  
Location           Cuisines                                               
Cedar Hill, NY     Indian       2.000000        5.000000   

Ao fazer agrupamentos, conseguimos identificar qual a média das culinárias no Central Park.

# Lidando com dados faltantes

In [11]:
house_SP = pd.read_csv('houses_sp.csv')
house_SP.head()

city   area  rooms  bathroom  parking spaces floor   hoa  rent   tax  \
0  São Paulo   70.0    2.0         1             1.0     7  2065  3300   211   
1  São Paulo  320.0    4.0         4             2.0    20  1200  4960  1750   
2  São Paulo   25.0    1.0         1             NaN     1     0   800    25   
3  São Paulo  650.0    3.0         3             7.0     -     0  8000   834   
4  São Paulo  213.0    NaN         4             4.0     4  2254  3223  1735   

   insurance   total  page hits  days available  interactions  \
0         42  5618.0        324              23           108   
1         63  7973.0        720              78           240   
2         11   836.0       1548              78           516   
3        121  8955.0        396              66           132   
4         41  7253.0        756              99           252   

   weeks available   type  
0                3   flat  
1               11   flat  
2               11   flat  
3                9  house  
4               14   flat

In [12]:
house_SP['city'].value_counts() 

city
São Paulo    7143
Name: count, dtype: int64

*Dado Constante*: Esta informação é irrelevante  uma vez que ela está presente em todo o dataset e também por se tratar de um banco de dados da cidade de SP, potanto é interessante remover este tipo de dado. 

In [13]:
house_SP = house_SP.drop(columns=['city'])
house_SP

area  rooms  bathroom  parking spaces floor   hoa   rent   tax  \
0      70.0    2.0         1             1.0     7  2065   3300   211   
1     320.0    4.0         4             2.0    20  1200   4960  1750   
2      25.0    1.0         1             NaN     1     0    800    25   
3     650.0    3.0         3             7.0     -     0   8000   834   
4     213.0    NaN         4             4.0     4  2254   3223  1735   
...     ...    ...       ...             ...   ...   ...    ...   ...   
7138   24.0    2.0         2             1.0    13   993   5500   141   
7139  280.0    4.0         4             2.0     5  4200   4000  1042   
7140   83.0    3.0         2             2.0    11   888   7521   221   
7141  150.0    3.0         3             2.0     8     0  13500     0   
7142  285.0    4.0         4             4.0    17  3100  15000   973   

      insurance    total  page hits  days available  interactions  \
0            42   5618.0        324              23           108   
1            63   7973.0        720              78           240   
2            11    836.0       1548              78           516   
3           121   8955.0        396              66           132   
4            41   7253.0        756              99           252   
...         ...      ...        ...             ...           ...   
7138         70   6704.0       1800              93           600   
7139         51   9293.0       1116              56           372   
7140         96   8726.0       1116              84           372   
7141        172  13672.0       2124             114           708   
7142        191  19264.0        612              99           204   

      weeks available   type  
0                   3   flat  
1                  11   flat  
2                  11   flat  
3                   9  house  
4                  14   flat  
...               ...    ...  
7138               13   flat  
7139                8   flat  
7140               12   flat  
7141               16   flat  
7142               14   flat  

[7143 rows x 15 columns]

Dados Null: Muitas vezes temos dados que foram inseridos incorretos ou sao faltantes. 

In [14]:
house_SP['rooms']

0       2.0
1       4.0
2       1.0
3       3.0
4       NaN
       ... 
7138    2.0
7139    4.0
7140    3.0
7141    3.0
7142    4.0
Name: rooms, Length: 7143, dtype: float64

A nossa database tem ao todo 7143 entradas e para o caso de *roons* temos 7143-7092 = 51 dados Nulls

In [15]:
house_SP.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7143 entries, 0 to 7142
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   area             7108 non-null   float64
 1   rooms            7092 non-null   float64
 2   bathroom         7143 non-null   int64  
 3   parking spaces   5586 non-null   float64
 4   floor            7143 non-null   object 
 5   hoa              7143 non-null   int64  
 6   rent             7143 non-null   int64  
 7   tax              7143 non-null   object 
 8   insurance        7143 non-null   int64  
 9   total            7143 non-null   float64
 10  page hits        7143 non-null   int64  
 11  days available   7143 non-null   int64  
 12  interactions     7143 non-null   int64  
 13  weeks available  7143 non-null   int64  
 14  type             7143 non-null   object 
dtypes: float64(4), int64(8), object(3)
memory usage: 837.2+ KB


1. A primeira alternativa é remover as linhas que contem os Nulls, contudo, esta solução remove os dados. 
2. A segunda opçao é escolher uma caracterizaçao estatística

In [16]:
house_SP['rooms'].median()

np.float64(3.0)

In [17]:
house_SP['rooms'] = house_SP['rooms'].fillna(house_SP['rooms'].median())

In [18]:
house_SP

area  rooms  bathroom  parking spaces floor   hoa   rent   tax  \
0      70.0    2.0         1             1.0     7  2065   3300   211   
1     320.0    4.0         4             2.0    20  1200   4960  1750   
2      25.0    1.0         1             NaN     1     0    800    25   
3     650.0    3.0         3             7.0     -     0   8000   834   
4     213.0    3.0         4             4.0     4  2254   3223  1735   
...     ...    ...       ...             ...   ...   ...    ...   ...   
7138   24.0    2.0         2             1.0    13   993   5500   141   
7139  280.0    4.0         4             2.0     5  4200   4000  1042   
7140   83.0    3.0         2             2.0    11   888   7521   221   
7141  150.0    3.0         3             2.0     8     0  13500     0   
7142  285.0    4.0         4             4.0    17  3100  15000   973   

      insurance    total  page hits  days available  interactions  \
0            42   5618.0        324              23           108   
1            63   7973.0        720              78           240   
2            11    836.0       1548              78           516   
3           121   8955.0        396              66           132   
4            41   7253.0        756              99           252   
...         ...      ...        ...             ...           ...   
7138         70   6704.0       1800              93           600   
7139         51   9293.0       1116              56           372   
7140         96   8726.0       1116              84           372   
7141        172  13672.0       2124             114           708   
7142        191  19264.0        612              99           204   

      weeks available   type  
0                   3   flat  
1                  11   flat  
2                  11   flat  
3                   9  house  
4                  14   flat  
...               ...    ...  
7138               13   flat  
7139                8   flat  
7140               12   flat  
7141               16   flat  
7142               14   flat  

[7143 rows x 15 columns]

In [19]:
house_SP.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7143 entries, 0 to 7142
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   area             7108 non-null   float64
 1   rooms            7143 non-null   float64
 2   bathroom         7143 non-null   int64  
 3   parking spaces   5586 non-null   float64
 4   floor            7143 non-null   object 
 5   hoa              7143 non-null   int64  
 6   rent             7143 non-null   int64  
 7   tax              7143 non-null   object 
 8   insurance        7143 non-null   int64  
 9   total            7143 non-null   float64
 10  page hits        7143 non-null   int64  
 11  days available   7143 non-null   int64  
 12  interactions     7143 non-null   int64  
 13  weeks available  7143 non-null   int64  
 14  type             7143 non-null   object 
dtypes: float64(4), int64(8), object(3)
memory usage: 837.2+ KB


Dados digitados errados

In [20]:
house_SP.head(10)

area  rooms  bathroom  parking spaces floor   hoa   rent   tax  insurance  \
0   70.0    2.0         1             1.0     7  2065   3300   211         42   
1  320.0    4.0         4             2.0    20  1200   4960  1750         63   
2   25.0    1.0         1             NaN     1     0    800    25         11   
3  650.0    3.0         3             7.0     -     0   8000   834        121   
4  213.0    3.0         4             4.0     4  2254   3223  1735         41   
5  152.0    2.0         2             1.0     3  1000  15000   250        191   
6   26.0    1.0         1             NaN     2   470   2100   150         27   
7   36.0    1.0         1             NaN    11   359   2100    70         27   
8   55.0    1.0         1             1.0     2   790   4200   224         54   
9  100.0    2.0         2             2.0    24   900   4370    17         56   

     total  page hits  days available  interactions  weeks available   type  
0   5618.0        324              23           108                3   flat  
1   7973.0        720              78           240               11   flat  
2    836.0       1548              78           516               11   flat  
3   8955.0        396              66           132                9  house  
4   7253.0        756              99           252               14   flat  
5  16441.0       2412             142           804               20   flat  
6   2747.0        828              85           276               12   flat  
7   2556.0        360              22           120                3   flat  
8   5268.0        540             120           180               17   flat  
9   5343.0       1044              65           348                9   flat

O data floor nos diz o andar do apartamentom, contudo quando estamos falando de um aparatamento, não temos um andar.
Podemos substituit o traço por um número condizendo com os dados. 

In [21]:
house_SP['floor'].value_counts() # A indidência é muito alta, não podemos apenas remover

floor
-      1847
1       727
2       506
4       402
3       399
5       370
7       312
6       297
8       291
10      256
11      241
9       230
14      206
13      206
12      190
15      120
16       84
17       71
21       58
20       56
18       56
26       54
19       45
25       36
22       22
23       21
24       18
27        7
28        6
29        4
32        2
35        1
110       1
46        1
Name: count, dtype: int64

Vamos definir qua a casa está no andar 0, contudo, está modificação precisa ser validada com a área de negocios.

In [22]:
house_SP.loc[house_SP['floor'] == '-', 'floor'] = '0'

In [23]:
house_SP.head(10)

area  rooms  bathroom  parking spaces floor   hoa   rent   tax  insurance  \
0   70.0    2.0         1             1.0     7  2065   3300   211         42   
1  320.0    4.0         4             2.0    20  1200   4960  1750         63   
2   25.0    1.0         1             NaN     1     0    800    25         11   
3  650.0    3.0         3             7.0     0     0   8000   834        121   
4  213.0    3.0         4             4.0     4  2254   3223  1735         41   
5  152.0    2.0         2             1.0     3  1000  15000   250        191   
6   26.0    1.0         1             NaN     2   470   2100   150         27   
7   36.0    1.0         1             NaN    11   359   2100    70         27   
8   55.0    1.0         1             1.0     2   790   4200   224         54   
9  100.0    2.0         2             2.0    24   900   4370    17         56   

     total  page hits  days available  interactions  weeks available   type  
0   5618.0        324              23           108                3   flat  
1   7973.0        720              78           240               11   flat  
2    836.0       1548              78           516               11   flat  
3   8955.0        396              66           132                9  house  
4   7253.0        756              99           252               14   flat  
5  16441.0       2412             142           804               20   flat  
6   2747.0        828              85           276               12   flat  
7   2556.0        360              22           120                3   flat  
8   5268.0        540             120           180               17   flat  
9   5343.0       1044              65           348                9   flat

Outro dado que pode gerar problemas está relacionado ao *parking spaces*, alguns estão contendo valores como *NaN*
Provavelmente indicando que as residencias não possuem estacionamentos. 
Vamos categorizar por tamanho. Obter a média de estácionamento para uma casa até uma certa área. Com isso indicamos que quanto maior a residência, mais estácionamento ela pode ter. 

## Casa

In [24]:
house_SP[(house_SP['type'] == 'house') & (house_SP['area'] <= 50)]['parking spaces'].mean()

np.float64(1.0869565217391304)

In [25]:
median_house_50  = house_SP[(house_SP['type'] == 'house') & (house_SP['area'] <= 50)]['parking spaces'].median()

Para casa até $50\,m^2$ tanto a médiana quanto a média é 1. Portanto vamos estabelecer o valor 1 para estas residências

In [26]:
house_SP.loc[(house_SP['type'] == 'house') & (house_SP['area'] <= 50) & (house_SP['parking spaces'].isna()), 'parking spaces'] = median_house_50

In [27]:
house_SP[(house_SP['type'] == 'house') & (house_SP['area'] <= 50) & (house_SP['parking spaces'])]['parking spaces'].isna().value_counts()

parking spaces
False    254
Name: count, dtype: int64

Podemos ver que agora temos apenas casas com valores. 

Vamos fazer o mesmo para casas entre $50\,m^2$ e $100\,m^2$

In [28]:
house_SP[(house_SP['type'] == 'house') & (house_SP['area'] >= 50) & (house_SP['area'] <= 100)]['parking spaces'].mean()

np.float64(1.5594713656387664)

In [29]:
median_house_50_100 = house_SP[(house_SP['type'] == 'house') & (house_SP['area'] >= 50) & (house_SP['area'] <= 100)]['parking spaces'].median()

In [30]:
house_SP.loc[(house_SP['type'] == 'house') & (house_SP['area'] >= 50) & (house_SP['area'] <= 100) & (house_SP['parking spaces'].isna()), 'parking spaces'] = median_house_50_100

In [31]:
house_SP[(house_SP['type'] == 'house') & (house_SP['area'] >= 50) & (house_SP['area'] <= 100)]['parking spaces'].isna().value_counts()

parking spaces
False    361
Name: count, dtype: int64

O mesmo para casas com mais de $100\,m^2$

In [32]:
house_SP[(house_SP['type'] == 'house') & (house_SP['area'] >= 100)]['parking spaces'].mean()

np.float64(3.4983870967741937)

In [33]:
median_house_greater_100 = house_SP[(house_SP['type'] == 'house') & (house_SP['area'] >= 100)]['parking spaces'].median()

In [34]:
house_SP.loc[(house_SP['type'] == 'house') & (house_SP['area'] >= 100) & (house_SP['parking spaces'].isna()), 'parking spaces'] = median_house_greater_100

In [35]:
house_SP[(house_SP['type'] == 'house') & (house_SP['area'] >= 100)]['parking spaces'].isna().value_counts()

parking spaces
False    1369
Name: count, dtype: int64

Agora todas as casas tiveram seus *parking spaces* preenchidos? 

In [36]:
house_SP['parking spaces'][(house_SP['type'] == 'house')].isna().value_counts()

parking spaces
False    1843
True        4
Name: count, dtype: int64

Ainda temos 4 linhas que possuem *parking spaces* como *NaN*

Olhando bem para estas 4 linhas, podemos ver que são 4 locais que não possuem área, por isso código anterior não pegou. Como é apenas 4 linhas, podemos remover.

In [37]:
house_SP[(house_SP['type'] == 'house') & (house_SP['parking spaces'])] = house_SP[(house_SP['type'] == 'house') & (house_SP['area'])].dropna()

In [38]:
house_SP['parking spaces'][(house_SP['type'] == 'house')].isna().value_counts()

parking spaces
False    1837
True        4
Name: count, dtype: int64

In [42]:
house_SP = house_SP.drop(house_SP[(house_SP['type'] == 'house') & (house_SP['parking spaces'].isna())].index)

Vamos fazer a mesma análise, mas agora para os apartamentos

## Apartamento